### ISSUE TITLE
Branch protection

### ISSUE THREAD

**Issue by asmeurer** [conda-forge/conda-smithy#255]
I want to enable branch protection in all my feedstocks (IMHO, it should be on by default for every feedstock). I can't do this because I don't have admin access. 

Preferably the protection would have all the checkboxes, possibly sans the one that require branches to be fast-forward for all the branches (which is just `master` by default, but since builds run on every branch, it should be applied to every new branch that is created too).

**Comment by asmeurer** [comment]
Ping

**Comment by asmeurer** [comment]
It looks like you can change branch protection from the [API](https://developer.github.com/v3/repos/branches/#update-branch-protection).

**Comment by jakirkham** [comment]
Sorry this has been quiet.

Honestly this is lower priority item for me. Higher priority items include handling `conda-build` version `2.0.0`, fixing periodic upload failures ( https://github.com/conda-forge/status/issues/4 ), getting some sort of matrix support on CircleCI ( https://github.com/conda-forge/conda-smithy/issues/203 ), and honestly fixing some AppVeyor oddities that are causing problems ( https://github.com/conda-forge/status/issues/9 ).

Though I'll try to respond to this recent point about the API. I'm not keen on forcing users into this functionality. If there is a way to make it optional, like adding some file to the repo or something else. I'd be in favor of doing that.

**Comment by asmeurer** [comment]
I'll be happy to take a shot at implementing this. Just want some questions answered:

- Why don't I have admin access on my feedstocks?
- Are you actually OK with enabling branch protection on all feedstocks?
- If so, what level? Is all levels OK (the highest level requires status checks to pass before merging, and even admin users can't override it)?

Keep in mind that conda-forge is in a perfect position to decide that this should be enabled if you want. It's a question of quality control (and given the pickiness of the linter, I think quality control is a high priority for conda-forge). Disabling force pushes makes it impossible to accidentally screw up the repo with an stray git command. Required status checks (a separate level) mean that the master of the feedstock always actually builds. I would not recommend the option that requires you to re-merge each PR when it is not fast-forward with master, as it just results in extra builds for the CI, and is unlikely to hide any real issues with repos such as feedstocks. The admin-only option is somewhat pointless (admins can just manually uncheck the box), but it's useful I guess to prevent admins from accidentally or late-nightetly clicking merge. 

And by the way, the easiest way to make it optional (and to make me shut up about it) is to give feedstock maintainers admin access to their repos. :wink: (although I really do think it would be advantageous to have this on for everyone)

**Comment by asmeurer** [comment]
For reference, these are the possible options for branch protection

<img width="769" alt="screen shot 2016-08-31 at 11 33 44 pm" src="https://cloud.githubusercontent.com/assets/71486/18154524/6b5c79ca-6fd3-11e6-966c-6ed11239ad8a.png">

**Comment by jakirkham** [comment]
> Why don't I have admin access on my feedstocks?

May have said this before. That's a question for @pelson. Though I can guess a few reasons.

> Are you actually OK with enabling branch protection on all feedstocks?

So was trying to answer this in my comment above, but I think it didn't come across.

I'm ok with providing feedstocks the option to set this. Right now I'm not seeing a clear path to accomplish this other than giving them admin privileges, which again it would be good to know the reasons not to do this.

> If so, what level? Is all levels OK (the highest level requires status checks to pass before merging, and even admin users can't override it)?

Not sure. Definitely strongly against the highest level.

There are practical reasons for one to merge a failing PR at conda-forge. Though I'm not ready to get into a long discussion about this at this time.

Personally I could get behind blocking force pushes. However as a community it seems there is still room to discuss this. See issue ( https://github.com/conda-forge/numpy-feedstock/issues/10 ). I had added this point to our agenda some time back, but don't know that it actually got resolved.

**Comment by pelson** [comment]
FWIW I'm personally 👍  on force protection and 👎 on status checks needing to pass.

The main reason that admin rights to the conda-forge feedstock repos only exist for conda-forge admins is to ensure that the feedstock remain loosely inline and consistent (from a settings point of view). For example, consistency on webhooks, branch protection, team management etc. is quite an important factor to enable people to quickly shift context between different feedstocks.

**Comment by jakirkham** [comment]
So sounds like we are ok with blocking force pushes across the board. If you want to try drumming up a PR like that, @asmeurer I think there is a good chance for it being accepted. Does that sound reasonable?

**Comment by jakirkham** [comment]
I'm starting to think that this would have to be a webservice. Though the functionality to make it work probably should still live in `conda-smithy`.

While we could add this to the `master` branch when creating a feedstock (and that probably satisfies most cases), this doesn't help us much when a new branch is created on a feedstock. One could run something locally when creating a new branch. However, they would need a GitHub token for it and the whole process gets a bit confusing for beginners (plus a bit much for core devs to explain).

OTOH if we have a webservice, we could listen for a [CreateEvent]( https://developer.github.com/v3/activity/events/types/#createevent ) to notice a new branch has been added to a feedstock. The webservice can then use a provided GitHub token to configure the right degree of protection for the new branch. Not sure if we will have to special case `master` or not, but we can think about that a bit more once there is code to enable some degree of branch protection for a branch.

In short, a webservice will basically be necessary to handle this for other branches added after conversion.

**Comment by asmeurer** [comment]
It would be nice if GitHub would let you set a default branch protection for a repository.

**Comment by jakirkham** [comment]
I suppose we could ask them to add this.
